### PR TITLE
dynamodbv2 : support nested maps with a single k/v pair

### LINKED ISSUE #120: dynamodbv2 cannot put item containing map with one key/value pair where key is one character long
When trying to put an item into a DynamoDB table which contains a map with just one key/value pair, where the key is one character long, one gets an `AmazonServiceException`, as the following example demonstrates:

```clj
(dynamo/put-item :table-name "rs2-dev-match"
                 :item {:id "m1", :foo {:a 1}})
```

The resulting stacktrace:

```
AmazonServiceException Supplied AttributeValue is empty, must contain exactly one of the supported datatypes (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: ValidationException; Request ID: DK3M098NKLDNV3LV85AQVHI2QVVV4KQNSO5AEMVJF66Q9ASUAAJG)
	com.amazonaws.http.AmazonHttpClient.handleErrorResponse (AmazonHttpClient.java:1077)
	com.amazonaws.http.AmazonHttpClient.executeOneRequest (AmazonHttpClient.java:725)
	com.amazonaws.http.AmazonHttpClient.executeHelper (AmazonHttpClient.java:460)
	com.amazonaws.http.AmazonHttpClient.execute (AmazonHttpClient.java:295)
	com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.invoke (AmazonDynamoDBClient.java:3106)
	com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient.putItem (AmazonDynamoDBClient.java:1206)
	sun.reflect.NativeMethodAccessorImpl.invoke0 (NativeMethodAccessorImpl.java:-2)
	sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:57)
	sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
	java.lang.reflect.Method.invoke (Method.java:606)
	sun.reflect.GeneratedMethodAccessor21.invoke (:-1)
	sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
```

The issue is the [`amazonica.aws.dynamodbv2/scalar-attribute-value?`](https://github.com/mcohen01/amazonica/blob/d9373fab246b592a981ddaa41e7b6daa22eb6225/src/amazonica/aws/dynamodbv2.clj#L38) function:

```clj
(defn- scalar-attribute-value? [value]
  (and (map? value)
       (= 1 (count value))
       (= 1 (count (name-of-first-key value)))))
```

@jankronquist, can you explain what the purpose of this function is? What is a scalar attribute? Are you looking for things like `{:S "foo"}` or `{:N "bar"}`? If so, this code won't handle `{:BOOL true}` and `{:SS ["foo" "bar"]}`, etc.

Am I misunderstanding the intent of the code?

### PR DESCRIPTION
**Pull request by jmglov**
Fixes #120

### CODE REVIEW

**Comment by mcohen01** [comment]
Thanks Josh

File: src/amazonica/aws/dynamodbv2.clj [hunk 1]
```diff
@@ -38,7 +38,7 @@
 (defn- scalar-attribute-value? [value]
   (and (map? value)
        (= 1 (count value))
-       (= 1 (count (name-of-first-key value)))))
+       (contains? #{:s :b :BOOL :SS :BS :n :NS :l :m :NULL} (name-of-first-key value))))
```
Comment (gws) [inline src/amazonica/aws/dynamodbv2.clj]: @jmglov Is this correct? `(name-of-first-key value)` should return a string, so it seems like the `contains?` check would never be true. I'm getting the following dynamodb-local error, new in amazonica 0.3.14:

```
com.amazonaws.AmazonServiceException: Invalid attribute value type (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: ValidationException; Request ID: 022030b9-6264-4174-a258-cd02d4ced32f)
```

and I believe this change is related. I can file a new issue but I thought I'd check first.

Comment (jmglov) [inline src/amazonica/aws/dynamodbv2.clj]: Good catch! I tested this with real Dynamo, and it worked, but you're
absolutely right that name-of-first-key returns a string.

I'll fix this and send a new pull request. If you tell me how to configure
Amazonica to use Dynamo Local, I'll be sure to test with that as well.

Please send me a code snippet that demonstrates this bug so I can be
completely sure.

On Saturday, 14 February 2015, Gordon Stratton <notifications@github.com>
wrote:

> In src/amazonica/aws/dynamodbv2.clj
> <https://github.com/mcohen01/amazonica/pull/121#discussion_r24705465>:
>
> > @@ -38,7 +38,7 @@
> >  (defn- scalar-attribute-value? [value]
> >    (and (map? value)
> >         (= 1 (count value))
> > -       (= 1 (count (name-of-first-key value)))))
> > +       (contains? #{:s :b :BOOL :SS :BS :n :NS :l :m :NULL} (name-of-first-key value))))
>
> @jmglov <https://github.com/jmglov> Is this correct? (name-of-first-key
> value) should return a string, so it seems like the contains? check would
> never be true. I'm getting the following dynamodb-local error, new in
> amazonica 0.3.14:
>
> com.amazonaws.AmazonServiceException: Invalid attribute value type (Service: AmazonDynamoDBv2; Status Code: 400; Error Code: ValidationException; Request ID: 022030b9-6264-4174-a258-cd02d4ced32f)
>
> and I believe this change is related. I can file a new issue but I thought
> I'd check first.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mcohen01/amazonica/pull/121/files#r24705465>.
>


-- 
Sent from my iPhone. Please excuse the brevity and potential typos.

Comment (gws) [inline src/amazonica/aws/dynamodbv2.clj]: Hey! All you need to reproduce is actually in the [README](https://github.com/mcohen01/amazonica#dynamodbv2) - the `:endpoint` key can take a URL as well as a region, so you just use `http://localhost:8000` or wherever DynamoDB-local is running.

The code snippet that demonstrates the bug is nearly identical to the `get-item` call that's demonstrated in that same section of the README - my case just had a single-entry map as a `:key` in the `get-table` call however, as shown below:

```clojure
(get-item cred :table-name "TestTable"
               :key {:id {:s "foo"}})
```

If you use Docker and want a really quick in-memory DynamoDB-local instance up and running on port 8000, you can try this:

```sh
docker run -it --rm -p 8000:8000 massiki/dynamodb-local
```

**Comment by jmglov** [comment]
@gws Fixed this in #122 

Thanks for bringing this to my attention, and sorry for causing you problems in the first place.

**Comment by gws** [comment]
@jmglov Thank you! It was no trouble, and I look forward to the next release.

**Pull request merged**